dartagnan/src/main/java/com/dat3m/dartagnan/encoding/PropertyEncoder.java

```diff
@@ -398,6 +402,93 @@ public TrackableFormula encodeDataRaces() {
+                .filter(MemoryObject::isHeapAllocated)
+                .collect(Collectors.toMap(o -> o, o -> new Var(context.leakVariable(o), context.trackVariable(o))));
+        // Requires all threads to exit normally.
+        for (Event event : program.getThreadEvents()) {
+            if (Stream.of(Tag.BOUND, Tag.EXCEPTIONAL_TERMINATION, Tag.NONTERMINATION).anyMatch(event::hasTag)) {
+                enc.add(bmgr.not(context.execution(event)));
+            }
+        }
+        // Do not leak unallocated objects.
+        for (Map.Entry<MemoryObject, Var> entry : variables.entrySet()) {
+            enc.add(bmgr.or(context.execution(entry.getKey().getAllocationSite()), bmgr.not(entry.getValue().leak)));
+        }
+        // Do not leak deallocated objects.
+        final ExpressionEncoder exprEncoder = context.getExpressionEncoder();
+        for (Dealloc dealloc : program.getThreadEvents(Dealloc.class)) {
+            final BooleanFormula deallocated = context.execution(dealloc);
+            final TypedFormula<?, ?> address = context.address(dealloc);
+            for (MemoryObject addressableObject : alias.addressableObjects(dealloc)) {
+                final Var addressVariable = variables.get(addressableObject);
+                final BooleanFormula addressed = exprEncoder.equal(address, context.address(addressableObject));
+                enc.add(bmgr.not(bmgr.and(deallocated, addressed, addressVariable.leak)));
+            }
+        }
+        // Object A is reachable from another object B, if a co-maximal store writes an address pointing to A into B.
+        for (Store store : program.getThreadEvents(Store.class)) {
+            if (!mayBeFinal(store)) { continue; }
+            final Set<MemoryObject> communicableObjects = alias.communicableObjects(store).stream()
+                    .filter(variables::containsKey)
+                    .collect(Collectors.toSet());
+            if (communicableObjects.isEmpty()) { continue; }
+            final BooleanFormula isFinal = context.lastCoVar(store);
+            for (MemoryObject addressableObject : alias.addressableObjects(store)) {
+                final Var addressVariable = variables.get(addressableObject);
+                if (addressVariable == null && !addressableObject.isStaticallyAllocated()) { continue; }
+                final BooleanFormula addressed = referencesObject(store, false, addressableObject);
+                final BooleanFormula addressTrack = addressVariable != null ? addressVariable.track : bmgr.makeTrue();
+                for (MemoryObject communicableObject : communicableObjects) {
+                    final Var valueVariable = variables.get(communicableObject);
+                    if (valueVariable == null) { continue; }
+                    final BooleanFormula valueTrack = valueVariable.track;
+                    final BooleanFormula communicated = bmgr.not(referencesObject(store, true, communicableObject));
```

Comment:
I don't understand the negation here? Doesn't it mean that the value of the store does NOT point into the object?